doc: Gatekeeper policies as OCI image blog (#81)
oras-project/oras-www: docs/blog/gatekeeper-policies-as-oci-image.md

> # Bundle with ORAS, test with `gator` and deploy with GitOps your Gatekeeper policies as OCI image

Reviewer: This title is a bit of a bag of words - maybe it's just me, but perhaps something shorter and more succinct?

Author: Good point, simplified it by `Bundle, test and deploy Gatekeeper policies as OCI image`

_Author: [Mathieu Benoit](https://www.linkedin.com/in/mathieubenoitqc/), DevRel Engineer, Google_

> [ORAS](https://oras.land/) is an important tool out there for working with OCI artifacts and OCI registries. As one of the users and advocates of ORAS, I witnessed the [growing trend in both user adoption and contributions in 2022](https://oras.land/blog/oras-looking-back-at-2022-and-forward-to-2023/). I thought I will share via this blog, an end-to-end scenario with OPA Gatekeeper policies and ORAS, from bundling to deploying them.

Reviewer: It would be appropriate to tell the reader what ORAS is - I can almost guarantee someone will come to the page and now know. E.g,. "ORAS "OCI Registry as Storage"

Author: Done

> [ORAS](https://oras.land/) is an important tool out there for working with OCI artifacts and OCI registries. As one of the users and advocates of ORAS, I witnessed the [growing trend in both user adoption and contributions in 2022](https://oras.land/blog/oras-looking-back-at-2022-and-forward-to-2023/). I thought I will share via this blog, an end-to-end scenario with OPA Gatekeeper policies and ORAS, from bundling to deploying them.

Reviewer: ```suggestion
[ORAS](https://oras.land/) is an important tool out there for working with OCI artifacts and OCI registries. As one of the users and advocates of ORAS, I witnessed the [growing trend in both user adoption and contributions in 2022](https://oras.land/blog/oras-looking-back-at-2022-and-forward-to-2023/). In this blog post I will share an end-to-end scenario with OPA Gatekeeper policies and ORAS, from including the steps from bundling to deployment.
```

Author: Done

> Policies are key to meet governance requirements as well as to improve the security of Kubernetes workloads and clusters. Policy engines like [OPA Gatekeeper](https://open-policy-agent.github.io/gatekeeper/website/docs/), [Kyverno](https://kyverno.io/) or even the new Kubernetes's [Validating Admission Policies](https://kubernetes.io/blog/2022/12/20/validating-admission-policies-alpha/) feature help write and enforce such policies. Once the policies are written, how to easily and securely share them with different projects and teams? How to deploy them across the fleet of clusters? How to evaluate them as early as possible in CI/CD pipelines?

Reviewer: ```suggestion
Policies are rules expressed in YAML that not only afford meeting governance requirements, but also improve the security of Kubernetes workloads and clusters. Policy engines like [OPA Gatekeeper](https://open-policy-agent.github.io/gatekeeper/website/docs/), [Kyverno](https://kyverno.io/) or even the new Kubernetes's [Validating Admission Policies](https://kubernetes.io/blog/2022/12/20/validating-admission-policies-alpha/) feature help write and enforce such policies. Once the policies are written, however, how do we easily and securely share them with different projects and teams? How do we deploy them across the fleet of clusters? How do we evaluate them as early as possible in CI/CD pipelines?
```

> In this blog we will demonstrate how to bundle and share Gatekeeper policies as an OCI image thanks to the [ORAS](https://oras.land/cli/) CLI, how to evaluate any Kubernetes manifests against this OCI image with the [`gator`](https://open-policy-agent.github.io/gatekeeper/website/docs/gator/) CLI and finally how to deploy this OCI image in Kubernetes clusters, in a GitOps way.

Reviewer: ```suggestion
In this blog post we will demonstrate how to bundle and share Gatekeeper policies as an OCI image using the [ORAS](https://oras.land/cli/) command line client, how to evaluate Kubernetes manifests against this OCI image with the [`gator`](https://open-policy-agent.github.io/gatekeeper/website/docs/gator/) command line client, and how to deploy this OCI image in Kubernetes clusters in a GitOps way.
```

Author: Done.

![Flow between ORAS, OCI registry, Gatekeeper and Config Sync](gatekeeper-policies-as-oci-image/overview-flow.png)

> _[Google Artifact Registry](https://cloud.google.com/artifact-registry) is used to illustrate this blog as the OCI registry, but you can take [any registries supporting OCI artifacts](https://oras.land/implementors/#registries-supporting-oci-artifacts). For the GitOps tool, we are using the OSS project: [Config Sync](https://github.com/GoogleContainerTools/kpt-config-sync), you can also use it as part of the [Anthos Config Management](https://cloud.google.com/anthos/config-management) service or even use other GitOps tools supporting OCI images like [FluxCD](https://fluxcd.io/flux/cheatsheets/oci-artifacts/)._

Reviewer: ```suggestion
_While we use the [Google Artifact Registry](https://cloud.google.com/artifact-registry) as the OCI registry for this example, you can use [any registry supporting OCI artifacts](https://oras.land/implementors/#registries-supporting-oci-artifacts). For the GitOps tool, we are using the OSS project: [Config Sync](https://github.com/GoogleContainerTools/kpt-config-sync), and you can also use it as part of the [Anthos Config Management](https://cloud.google.com/anthos/config-management) service or even use other GitOps tools supporting OCI images like [FluxCD](https://fluxcd.io/flux/cheatsheets/oci-artifacts/)._
```

Author: Done.

## Create a Gatekeeper policy

> Let's create a [Gatekeeper policy](https://open-policy-agent.github.io/gatekeeper/website/docs/howto) composed by one `Constraint` and one `ConstraintTemplate` which will be leveraged throughout this blog. In this example, we are making sure that any non-system namespaces is leveraging the [Pod Security Admission](https://kubernetes.io/docs/concepts/security/pod-security-admission/) feature by having the appropriate label. 

Reviewer: ```suggestion
Let's create a [Gatekeeper policy](https://open-policy-agent.github.io/gatekeeper/website/docs/howto) composed by one `Constraint` and one `ConstraintTemplate` that will be leveraged throughout this blog. In this example, we are making sure that any non-system namespaces is leveraging the [Pod Security Admission](https://kubernetes.io/docs/concepts/security/pod-security-admission/) feature by having the appropriate label. 
```

Author: Done.

Create a dedicated folder for the associated files:
```bash

> mkdir policies

Reviewer: ```suggestion
mkdir -p policies
```

           not re_match(expected.allowedRegex, value)
           msg := sprintf("Label %s: %s does not satisfy allowed regex: %s", [key, value, expected.allowedRegex])
        }
EOF
```

Define the associated `Constraint` for the `Namespaces` which must have the label `pod-security.kubernetes.io/enforce` as the key and either `baseline` or `restricted` as the value:
```bash
cat <<EOF> policies/ns-must-have-pss-label.yaml
apiVersion: constraints.gatekeeper.sh/v1beta1
kind: K8sRequiredLabels
metadata:
  name: ns-must-have-pss-label
spec:
  enforcementAction: deny
  match:
    kinds:
      - apiGroups:
        - ""
        kinds:
        - "Namespace"
    excludedNamespaces:
    - config-management-monitoring
    - config-management-system
    - gatekeeper-system
    - kube-node-lease
    - kube-public
    - kube-system
    - resource-group-system
  parameters:
    labels:
    - key: pod-security.kubernetes.io/enforce
      allowedRegex: (baseline|restricted)
EOF
```

## Test this policy with local files

Define a `Namespace` without the PSS label.
```bash
cat <<EOF > namespace-test.yaml
apiVersion: v1
kind: Namespace
metadata:
  name: test
EOF
```

> Let’s now locally test this `Namespace` against this policy with the `gator` CLI. Very convenient to test policies without any Kubernetes cluster!

Reviewer: ```suggestion
Let’s now locally test this `Namespace` against this policy with the `gator` CLI. This client makes it very convenient to test policies without any Kubernetes cluster!
```

```bash
gator test \
   --filename namespace-test.yaml \
   --filename policies/
```

Output similar to:
```plaintext
["ns-must-have-pss-label"] Message: "You must provide labels: {\"pod-security.kubernetes.io/enforce\"} for the Namespace: test."
```

## Push the Gatekeeper policy as OCI image to Artifact Registry

Assuming we already have a [Google Artifact Registry repository](https://cloud.google.com/artifact-registry/docs/repositories/create-repos) `ARTIFACT_REGISTRY_REPO_NAME` in the region `REGION` and project `PROJECT_ID`.

Login to Artifact Registry:
```bash
gcloud auth configure-docker ${REGION}-docker.pkg.dev
```

Push the Gatekeeper policy as OCI image in Google Artifact Registry repository with ORAS:
```bash
oras push \
    ${REGION}-docker.pkg.dev/${PROJECT_ID}/${ARTIFACT_REGISTRY_REPO_NAME}/my-policies:1.0.0 \

>     policies/

Reviewer 2: pushing without an artifactType or --config mediaType means the gatekeeper policy has no type. Or, the default of unknown

Suggest providing some type by adding:
`--config /dev/null:application/vnd.cncf.gatekeeper.v1`

Reviewer 2: cc: @garethr for the actual mediaType as I know he'd done some work earlier.
Another suggestion would be to include the oras-go library into `gator` so you could do a `gator push|pull` directly. In this case, you could embed the appropriate `artifactType` directly into the gather cli.

Author: Thanks for your review @SteveLasker!
Yes indeed the `config.mediaType` is ``:
```
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.oci.image.manifest.v1+json",
   "config": {
      "mediaType": "application/vnd.unknown.config.v1+json",
      "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
      "size": 2
   },
   "layers": [
      {
         "mediaType": "application/vnd.oci.image.layer.v1.tar+gzip",
         "digest": "sha256:325baa2355262c1cb5d8fb274759405f5c4b26e2110dfc22bbf9565740757938",
         "size": 1027,
         "annotations": {
            "io.deis.oras.content.digest": "sha256:6d2f0cfca84142481087d9566dd9e3a77a69d3e1a3754595c1615d09348487d8",
            "io.deis.oras.content.unpack": "true",
            "org.opencontainers.image.title": "policies"
         }
      }
   ],
   "annotations": {
      "org.opencontainers.image.created": "2023-02-08T17:11:52Z"
   }
}
```

Where this `mediaType` `application/vnd.cncf.gatekeeper.v1` is coming from and defined? Is it like a feature request you are suggesting for the Gatekeeper/`gator` project? Not seeing anything in that regard in there: https://open-policy-agent.github.io/gatekeeper/website/docs/gator/#bundling-policy-into-oci-artifacts.

Author: Regarding your other point about `oras-go` in `gator`, that's already the case: https://github.com/open-policy-agent/gatekeeper/pull/2398. That's what `gator test|expand --image` is leveraging, FYI.

Author: Also, @SteveLasker, is it blocking this PR to get your approval to be shipped in `main`? Maybe this should be a separate discussion in Slack or also, I'm more than happy to file an issue in Gatekeeper/`gator` repo to have this discussion there (depending on the answers of my questions above). Thanks!

Reviewer 2: Thanks, @mathieu-benoit 
The mediaType usage is outlined here: [OCI Artifacts: Defining a Unique Artifact Type](https://github.com/opencontainers/artifacts/blob/main/artifact-authors.md#defining-a-unique-artifact-type)

As for which `config.mediaType` to use, that should be defined by the OPA project, and that's why I was pinging @garethr, as I remember him doing so in a prior experiment. 

I wouldn't say it's blocking, rather the article does a great job and users tend to copy/paste examples, so I was hoping to add this clarity so everything in the registry didn't look "unknown". 
Here's another example [Visualizing Artifacts](https://github.com/opencontainers/artifacts/blob/main/artifact-authors.md#visualizing-artifacts), which is also supported by Docker Hub. See [Announcing Docker Hub OCI Artifacts Support](https://www.docker.com/blog/announcing-docker-hub-oci-artifacts-support/), with the visualization section title: **Viewing your Helm chart and using filters**

I see the additional gator support, which is awesome. It just looks like this `config.mediaType` was copied over as unknown, rather than identifying a well-known type.

Author: Gotcha, thanks for clarifying. 

I'll go ahead and resolve this discussion to unblock the review/shipment of this blog post. Hope that's ok?

As soon as the `config.mediaType` value to use is defined by the OPA/Gatekeeper project, it will be my pleasure to update that blog post accordingly.

Thanks!

```
_Note: `oras push` with the `policies/` folder will do an archive of its files._

See that the OCI image has been uploaded in the Google Artifact Registry repository:
```bash
gcloud artifacts docker images list ${REGION}-docker.pkg.dev/${PROJECT_ID}/${ARTIFACT_REGISTRY_REPO_NAME}
```

## Test this policy with the remote OCI image

> Let’s now locally test the `Namespace` created earlier against this policy as remote OCI image with the gator CLI. Very convenient to share and evaluate your policies in different places (i.e. locally, during Continuous Integration pipelines, etc.)!

Reviewer: ```suggestion
Let’s now locally test the `Namespace` created earlier against this policy as remote OCI image with the gator CLI. It is very convenient to share and evaluate your policies in different places (i.e. locally, during Continuous Integration pipelines, etc.)!
```
